### PR TITLE
Initial authentication/authorization module

### PR DESCRIPTION
**Pull request by rgwozdz**
Create an authentication module that:

* decorates a provider model prototype with three functions: `authenticationSpecification`, `authenticate`, and `validateToken`.
* allows users to pass in there own credential-checking method

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# Koop-Direct-Auth
+## A module for implementing direct authentication from client to Koop server
+
+This sample provider demonstrates the implementation of token-based security on services. The provider itself simply serves static GeoJSON files as Koop output-services. A more complete description of Koop and Koop providers is available at [koopjs.github.io](http://koopjs.github.io/).
```
Comment (dmfenton) [inline README.md]: This isn't a provider

File: README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+![enter image description here](https://gist.githubusercontent.com/rgwozdz/e44f3686abe40360532fbcc6dccf225d/raw/9768df32fc62e99ce7383c124cab8efdf45b1e18/koop-direct-auth-resources.png)
+
+## Example of Koop authentication implementation
+
+The [server.js](./server.js) file provides an example of securing a provider's resources. Start by requiring the provider and the authentication module.
+
+    const provider = require('./')
+
+    const auth = require('./koop-auth-direct')
```
Comment (dmfenton) [inline README.md]: I think the best thing to show here is how to use this from inside a provider's model? Agree?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    const validateCredentials = require('your-identity-module-here')
+
+Supply the auth module with required parameters and use the ArcGIS online strategy to set options. Get back a configured middleware.
+
+	auth(provider, <a-secret-for-encoding>, validateCredentials)
+
+## Authentication API
+
+### (provider, secret, validateCredentials, options) ⇒ <code>void</code>
+* Decorates a provider's model with configured versions of the "validateToken", "authenticate" and "authenticationSpecification" functions
```
Comment (dmfenton) [inline README.md]: Would this work instead with `provider.Model.Prototype.authenticate` and `provider.Model.Prototype.authorize`?

**Review by dmfenton — COMMENTED** [review]
Since this is now external to a provider, I think we should make it function more like other Koop plugins.

e.g.

const auth = require('koop-auth-direct')({options})
koop.register(auth)
=> registered auth plugin for all providers

File: README.md (hunk 3 above)
Comment (rgwozdz) [reply README.md]: In terms of previous discussion, the  `validateToken` function here is equivalent to what we were referring to as `authorize` - so I will change the prototype function name  from `validateToken` to `authorize`.  I think we still need an `authenticationSpecification` prototype function - it serves out information used by output-services to point clients to a URL for authentication.

File: README.md (hunk 2 above)
Comment (rgwozdz) [reply README.md]: Not sure on that.  Currently, I don't refer to any auth code in the provider's model.js file.  Functions get added to the Model.Prototype but they're called in output services.  I'll put in a PR for koop-ouput-services so you can get a visual.

**Comment by rgwozdz** [comment]
One higher-level question in response to the "registration" suggestion:
```
const auth = require('koop-auth-direct')({options})
koop.register(auth)
=> registered auth plugin for all providers
```

Do we want to have auth-plugin automatically applied to all registered providers?  Or do we want a more explicit registration that looks for a property on the Provider? (e.g. `auth: true`).

**Comment by rgwozdz** [comment]
Another higher level question.  Do we want to allow different identity-management approaches across providers?  For example, say you have a koop instance with two providers, and you want to have different sets of users authorized for each provider.  If we allow auth to be implemented differently across providers (e.g. different `validateCredentials` function per provider) we could support this use-case.

**Comment by rgwozdz** [comment]
@dmfenton - refactor as per discussion earlier today. 8718c2f

File: README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    koop.register(auth)
+
+Then require and register your providers.  
+
+    const provider = require('./')
+    koop.register(provider)
+
+The authentication module will configure and add its `authorize`, `authenticate`, and `authenticationSpecification` functions to the provider's model prototype.  Output services will leverage these functions to secure the service endpoints and properly route requests to authenticate.
+
+Finally, create a JSON file store.  This should be an array of objects with properties `username` and `password`.  Set an environment variable `USER_STORE` with the path of the file relative to the root of the repository (e.g, `USER_STORE=./user-store.json`)
```
Comment (dmfenton) [inline README.md:28]: why not pass in the user store as an argument when setting up the auth plugin?

Comment (rgwozdz) [reply README.md:28]: I don't have a specific reason not to, in fact, I think it's better that way, have it fail on start up if the file is not there. I think I was in the mindset of trying to simplify the API as much as possible, but I'm happy to add as an argument.

Comment (rgwozdz) [reply README.md:28]: Updated in c0208d5

**Review by dmfenton — APPROVED** [review]

**Pull request merged**